google/cloud/internal/backoff_policy.cc

```diff
@@ -42,14 +42,18 @@ std::chrono::milliseconds ExponentialBackoffPolicy::OnCompletion() {
     generator_ = google::cloud::internal::MakeDefaultPRNG();
   }
   std::uniform_int_distribution<microseconds::rep> rng_distribution(
-      current_delay_range_.count() / 2, current_delay_range_.count());
+      current_delay_start_.count(), current_delay_end_.count());
   // Randomized sleep period because it is possible that after some time all
   // client have same sleep period if we use only exponential backoff policy.
   auto delay = microseconds(rng_distribution(*generator_));
-  current_delay_range_ = microseconds(static_cast<microseconds::rep>(
-      static_cast<double>(current_delay_range_.count()) * scaling_));
-  if (current_delay_range_ >= maximum_delay_) {
-    current_delay_range_ = maximum_delay_;
+
+  current_delay_start_ = current_delay_start_ == current_delay_end_
+                             ? initial_delay_
+                             : current_delay_end_;
+  current_delay_end_ = microseconds(static_cast<microseconds::rep>(
+      static_cast<double>(current_delay_end_.count()) * scaling_));
+  if (current_delay_end_ >= maximum_delay_) {
+    current_delay_end_ = maximum_delay_;
```

Comment:
Opt Nit: Perhaps use `std::min()` here?
```
current_delay_end_ = std::min(f(current_delay_end_), maximum_delay_);
```

Reply:
Nit: Notice also how the initialization of `current_delay_end_` (from `initial_delay_`) is different from its re-initialization (from itself).  The latter is careful to use floating-point multiplication, while the former is not.

It **may** make things easier if `current_delay_start_` and `current_delay_end_` were of type `std::chrono::duration<std::micro, double>` instead.  Then we could just say `current_delay_end_ * scaling_`.

Reply:
I also considered that, but with all of the casting I felt the original intention was obscured away

Alternatively changing the types requires casting when using  casting elsewhere since it is not compatible with `initial_delay_`

Reply:
Consider these diffs.  I believe that clarify the intention, and reduce casting.
```
diff --git a/google/cloud/internal/backoff_policy.cc b/google/cloud/internal/backoff_policy.cc
index 61bbe4df56..d4c21acd00 100644
--- a/google/cloud/internal/backoff_policy.cc
+++ b/google/cloud/internal/backoff_policy.cc
@@ -25,9 +25,6 @@ std::unique_ptr<BackoffPolicy> ExponentialBackoffPolicy::clone() const {
 }
 
 std::chrono::milliseconds ExponentialBackoffPolicy::OnCompletion() {
-  using std::chrono::duration_cast;
-  using std::chrono::microseconds;
-  using std::chrono::milliseconds;
   // We do not want to copy the seed in `clone()` because then all operations
   // will have the same sequence of backoffs. Nor do we want to use a shared
   // PRNG because that would require locking and some more complicated lifecycle
@@ -41,21 +38,17 @@ std::chrono::milliseconds ExponentialBackoffPolicy::OnCompletion() {
   if (!generator_) {
     generator_ = google::cloud::internal::MakeDefaultPRNG();
   }
-  std::uniform_int_distribution<microseconds::rep> rng_distribution(
+  std::uniform_real_distribution<DoubleMicroseconds::rep> rng_distribution(
       current_delay_start_.count(), current_delay_end_.count());
+
   // Randomized sleep period because it is possible that after some time all
   // client have same sleep period if we use only exponential backoff policy.
-  auto delay = microseconds(rng_distribution(*generator_));
+  auto delay = DoubleMicroseconds(rng_distribution(*generator_));
 
-  current_delay_start_ = current_delay_start_ == current_delay_end_
-                             ? initial_delay_
-                             : current_delay_end_;
-  current_delay_end_ = microseconds(static_cast<microseconds::rep>(
-      static_cast<double>(current_delay_end_.count()) * scaling_));
-  if (current_delay_end_ >= maximum_delay_) {
-    current_delay_end_ = maximum_delay_;
-  }
-  return duration_cast<milliseconds>(delay);
+  current_delay_start_ = current_delay_end_;
+  current_delay_end_ = CurrentDelayEnd();
+
+  return std::chrono::duration_cast<std::chrono::milliseconds>(delay);
 }
 
 }  // namespace internal
diff --git a/google/cloud/internal/backoff_policy.h b/google/cloud/internal/backoff_policy.h
index 19b4ddebde..b5721ec7c7 100644
--- a/google/cloud/internal/backoff_policy.h
+++ b/google/cloud/internal/backoff_policy.h
@@ -132,10 +132,7 @@ class ExponentialBackoffPolicy : public BackoffPolicy {
             maximum_delay)),
         scaling_(scaling),
         current_delay_start_(initial_delay_),
-        current_delay_end_(
-            (std::min)(std::chrono::duration_cast<std::chrono::microseconds>(
-                           scaling_ * initial_delay_),
-                       maximum_delay_)) {
+        current_delay_end_(CurrentDelayEnd()) {
     if (scaling_ <= 1.0) {
       google::cloud::internal::ThrowInvalidArgument(
           "scaling factor must be > 1.0");
@@ -157,12 +154,22 @@ class ExponentialBackoffPolicy : public BackoffPolicy {
   std::chrono::milliseconds OnCompletion() override;
 
  private:
+  using DoubleMicroseconds =
+      std::chrono::duration<double, std::chrono::milliseconds::period>;
+
+  DoubleMicroseconds CurrentDelayEnd() {
+    return std::min<DoubleMicroseconds>(current_delay_start_ * scaling_,
+                                        maximum_delay_);
+  }
+
   std::chrono::microseconds initial_delay_;
   std::chrono::microseconds maximum_delay_;
   double scaling_;
+
   // Stores both ends of the current delay range.
-  std::chrono::microseconds current_delay_start_;
-  std::chrono::microseconds current_delay_end_;
+  DoubleMicroseconds current_delay_start_;
+  DoubleMicroseconds current_delay_end_;
+
   absl::optional<DefaultPRNG> generator_;
 };
 
```